File: backend/src/main/scala/com/skewwhiffy/auraltester/controller/ScaleController.scala
package com.skewwhiffy.auraltester.controller

import com.skewwhiffy.auraltester.controller.dto.ScaleResponse
import com.skewwhiffy.auraltester.internalnotation.InternalNotationFactory
import com.skewwhiffy.auraltester.scales.{Key, ScaleDirection, ScaleTypeFactory}
import com.skewwhiffy.auraltester.services.{AbcService, ScaleService}
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.{GetMapping, RequestMapping, RequestParam, RestController}

import scala.util.chaining.scalaUtilChainingOps

@RestController
@RequestMapping(path = Array("/api/scale"))
class ScaleController(
  @Autowired private val abcService: AbcService,
  @Autowired private val internalNotationFactory: InternalNotationFactory,
  @Autowired private val scaleService: ScaleService,
  @Autowired private val scaleTypeFactory: ScaleTypeFactory
) {
  @GetMapping
  def get(
    @RequestParam(required = true) clef: String,
    @RequestParam(required = true) note: String,
    @RequestParam(required = true) scaleType: String,
    @RequestParam(required = true) direction: String
  ): ScaleResponse = {
    val clefObject = internalNotationFactory.clef(clef)
    val noteObject = internalNotationFactory.getNote(note).note
    val scaleTypeObject = scaleType match {
      case "major" => scaleTypeFactory.major
      case "minor-harmonic" => scaleTypeFactory.minorHarmonic
      case "minor-melodic" => scaleTypeFactory.minorMelodic
      case _ => throw new IllegalArgumentException(s"Unrecognized scale type: '$scaleType'")
    }
    val isMinor = scaleTypeObject.displayName != "major"
    val directionObject = direction match {
      case "ascending" => ScaleDirection.ascending
      case "descending" => ScaleDirection.descending
      case _ => throw new IllegalArgumentException(s"Unrecognized direction: '$direction'")
    }
    val scale = scaleService.getScale(clefObject, noteObject, scaleTypeObject, directionObject)
    val key = scale.lowestNote.note.pipe(it => new Key(it, isMinor))
    new ScaleResponse(abcService.getAbc(clefObject, scale), abcService.getAbc(clefObject, scale, key))
  }
}

File: backend/src/main/kotlin/com/skewwhiffy/controller/ScaleController.kt
package com.skewwhiffy.controller

import com.skewwhiffy.dto.ScaleResponse
import com.skewwhiffy.notation.factory.InternalNotationFactory
import com.skewwhiffy.notation.factory.ScaleTypeFactory
import com.skewwhiffy.notation.model.Key
import com.skewwhiffy.notation.model.scale.ScaleDirection
import com.skewwhiffy.service.AbcService
import com.skewwhiffy.service.ScaleService
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
class ScaleController(
  private val abcService: AbcService,
  private val internalNotationFactory: InternalNotationFactory,
  private val scaleService: ScaleService,
  private val scaleTypeFactory: ScaleTypeFactory,
) {
  @RequestMapping("/api/scale")
  fun get(
    clef: String,
    note: String,
    scaleType: String,
    direction: String,
  ): ScaleResponse {
    val clefObject = internalNotationFactory.clef(clef)
    val noteObject = internalNotationFactory.getNote(note).note
    val (scaleTypeObject, isMinor) = when (scaleType) {
      "major" -> Pair(scaleTypeFactory.major, false)
      "minor-harmonic" -> Pair(scaleTypeFactory.minorHarmonic, true)
      "minor-melodic" -> Pair(scaleTypeFactory.minorMelodic, true)
      else -> throw IllegalArgumentException("Unrecognized scale type: '$scaleType'")
    }
    val directionObject = when (direction) {
      "ascending" -> ScaleDirection.ASCENDING
      "descending" -> ScaleDirection.DESCENDING
      else -> throw IllegalArgumentException("Unrecognized direction: '$direction'")
    }
    val scale = scaleService.getScale(clefObject, noteObject, scaleTypeObject, directionObject)
    val key = scale.lowestNote.note.let { Key(it, isMinor) }
    return ScaleResponse(
      abcService.getAbc(clefObject, scale),
      abcService.getAbc(clefObject, scale, key)
    )
  }
}
